src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+            }
+        }
+
+        const scopes: Scope[] = [];
+        while (current) {
+            // We want to find the nearest parent where we can place an "equivalent" sibling to the node we're extracting out of.
+            // Walk up to the closest parent of a place where we can logically put a sibling:
+            //  * Function declaration
+            //  * Class declaration or expression
+            //  * Module or source file
+            // Note that we don't use isFunctionLike because we don't want to put the extracted closure *inside* a method
+            if ((current.kind === SyntaxKind.FunctionDeclaration) || isSourceFile(current) || isModuleBlock(current) || isClassLike(current)) {
+                scopes.push(current as FunctionLikeDeclaration);
+            }
+            current = current.parent;
+        }
+        return scopes;
+    }
+
+    /**
+     * Given a piece of text to extract ('targetRange'), computes a list of possible extractions.
+     * Each returned ExtractResultForScope corresponds to a possible target scope and is either a set of changes
+     * or an error explaining why we can't extract into that scope.
+     */
+    export function extractRange(targetRange: TargetRange, sourceFile: SourceFile, context: RefactorContext): ReadonlyArray<ExtractResultForScope> {
+        if (targetRange === undefined) {
+            return [];
+        }
+
+        const scopes = collectEnclosingScopes(targetRange);
+        if (scopes.length === 0) {
+            return [];
+        }
+        const enclosingTextRange = getEnclosingTextRange(targetRange, sourceFile);
+        const { target, usagesPerScope, errorsPerScope } = collectReadsAndWrites(
+            targetRange,
+            scopes,
+            enclosingTextRange,
+            sourceFile,
+            context.program.getTypeChecker());
+
+        context.cancellationToken.throwIfCancellationRequested();
+        return scopes.map((scope, i) => {
+            const errors = errorsPerScope[i];
+            if (errors.length) {
+                return {
+                    scope,
+                    scopeDescription: getDescriptionForScope(scope),
+                    errors
+                };
+            }
+            return extractFunctionInScope(target, scope, usagesPerScope[i], targetRange, context);
+        });
+    }
+
+    function getDescriptionForScope(s: Scope) {
+        if (isFunctionLike(s)) {
+            switch (s.kind) {
+                case SyntaxKind.Constructor:
+                    return "constructor";
+                case SyntaxKind.FunctionExpression:
+                    return s.name
+                        ? `function expression ${s.name.getText()}`
+                        : "anonymous function expression";
+                case SyntaxKind.FunctionDeclaration:
+                    return `function ${s.name.getText()}`;
+                case SyntaxKind.ArrowFunction:
+                    return "arrow function";

```

Comment:
I think Roslyn decided not to localize syntax kind names, but their use case is slightly different.